File: sendfile-Attachments/contracts-java/src/main/java/net/corda/examples/sendfile/contracts/InvoiceContract.java
package net.corda.examples.sendfile.contracts;

import net.corda.core.contracts.CommandData;
import net.corda.core.contracts.CommandWithParties;
import net.corda.core.contracts.Contract;
import net.corda.core.transactions.LedgerTransaction;
import net.corda.examples.sendfile.states.InvoiceState;
import org.jetbrains.annotations.NotNull;

import java.util.List;

import static net.corda.core.contracts.ContractsDSL.*;
// ************
// * Contract *
// ************
public class InvoiceContract implements Contract {
    // Used to identify our contract when building a transaction.
    public final static String ID = "net.corda.examples.sendfile.contracts.InvoiceContract";

    @Override
    public void verify(@NotNull LedgerTransaction tx) throws IllegalArgumentException {
        CommandWithParties<Commands> commandData = requireSingleCommand(tx.getCommands(), Commands.class);

        if (commandData.getValue() instanceof Commands.Issue) {
            requireThat(req -> {
                List<InvoiceState> output = tx.outputsOfType(InvoiceState.class);
                req.using("must be single output", output.size() == 1);
                req.using("Attachment ID must be stored in state", !output.get(0).getInvoiceAttachmentID().isEmpty());
                return null;
            });
        }
    }

    public interface Commands extends CommandData {
        class Issue implements Commands {}
    }
}

File: sendfile-Attachments/contracts-kotlin/src/main/kotlin/net/corda/examples/sendfile/contracts/InvoiceContract.kt
package net.corda.examples.sendfile.contracts

import net.corda.examples.sendfile.states.InvoiceState
import net.corda.core.contracts.CommandData
import net.corda.core.contracts.Contract
import net.corda.core.contracts.requireSingleCommand
import net.corda.core.contracts.requireThat
import net.corda.core.transactions.LedgerTransaction

// ************
// * Contract *
// ************
class InvoiceContract : Contract {
    companion object {
        // Used to identify our contract when building a transaction.
        const val ID = "net.corda.examples.sendfile.contracts.InvoiceContract"
    }

    // A transaction is valid if the verify() function of the contract of all the transaction's input and output states
    // does not throw an exception.
    override fun verify(tx: LedgerTransaction) {
        // Verification logic goes here.
        val cmd = tx.commands.requireSingleCommand<Commands>()
        when(cmd.value){
            is Commands.Issue -> requireThat {
                val output = tx.outputsOfType<InvoiceState>().single()
                "Attachment ID must be stored in state" using (output.invoiceAttachmentID.isNotEmpty())
            }
        }

    }

    // Used to indicate the transaction's intent.
    interface Commands : CommandData {
        class Issue : Commands
    }
}